rfcs/0008-react-layout.md

```diff
@@ -0,0 +1,201 @@
+- Start Date: 2020-01-28
+- Project: `@moxy/react-layout`
+- Reference Issues:
+- Implementation PR:
+
+# Summary
+
+Ability to define which layout to use per page.
+
+# Basic example
+
+Setup `<LayoutManager>` in your `App.js` component:
+
+```js
+import { LayoutManager } from '@moxy/react-layout';
+
+// Example Next.js App
+const App = ({ Component, pageProps }) => (
+    <LayoutManager>
+        <Component { ...pageProps } />
+    </LayoutManager>
+);
+```
+
+...and then use `withLayout` in your page components, e.g.: in `About.js`:
+
+```js
+import React from 'react';
+import { withLayout } from '@moxy/react-layout';
+import { PrimaryLayout } from '../../shared/components';
+import styles from './About.module.css';
+
+const About = () => (
+    <div className={ styles.about }>
+        <h1>About</h1>
+    </div>
+);
+
+export default withLayout(<PrimaryLayout variant="light" />)(About);
+```
+
+# Motivation
+
+Frontend projects usually have the need to have one or more layouts. Layouts are the "shell" of your app and usually contain navigation elements, such as an header and a footer. In the ideal scenario, each page would be able to say which layout they want to use, including tweaking its properties dynamically, such as `variant="light"`. However, we also want to keep the layout persistent in the React tree, to avoid having to remount it every time a user navigate between pages.
+
+Historically, projects overlook the need of multiple layouts or the ability to change layout props between pages. They start off with a simple layout and only later they handle this need, often with poor and non-scalable solutions.
+
+I propose to create a library, called `@moxy/react-layout`, that solves the need for multi-layouts and changing layout props dynamically in a consistent and reusable way.
+
+# Detailed design
+
+`@moxy/react-layout` is a very simple library that exposes a `<LayoutManager>` component and a `withLayout` to wrap pages.
+
+## &lt;LayoutManager&gt;
+
+A component that selects the layout based on what pages specify. It keeps the layout persistent between page transitions whenever possible (e.g.: when the layout is the same).
+
+Here's the list of props it supports:
+
+### defaultLayout
+
+Type: `ReactElement`
+
+The default layout to be used when a child page doesn't explicitly set one.
+
+```js
+// App.js
+import { LayoutManager } from '@moxy/react-layout';
+import { PrimaryLayout } from '../shared/components';
+
+const App = ({ Component, pageProps }) => (
+    <LayoutManager defaultLayout={ <PrimaryLayout /> }>
+        <Component { ...pageProps } />
+    </LayoutManager>
+);
+```
+
+### children
```

Comment:
Instead of using `render` prop, couldn't `children` either accept a `ReactComponent` or a `function`?

```js
<LayoutManager
   defaultLayout={ <PrimaryLayout /> }>
   { ({ Layout, layoutProps, layoutKey, pageKey }) => {
      ...
   } }
</LayoutManager>

Reply:
Applied your suggestion!